p2p/addressbook/store.go

```diff
@@ -58,19 +64,53 @@ func (a *AddrBook) persistPeers(path string) {
 		}
 	}
 
-	w, err := os.Create(path)
+	//w, err := os.Create(path)
+	//if err != nil {
+	//	a.logger.Error("Error creating file: %v", err)
+	//	return
+	//}
+	//enc := json.NewEncoder(w)
+	//defer w.Close()
+	//if err = enc.Encode(&sam); err != nil {
+	//	a.logger.Error("Failed to encode file %s: %v", a.path, err)
+	//	return
+	//}
+	data, err := json.Marshal(sam)
 	if err != nil {
-		a.logger.Error("Error creating file: %v", err)
+		a.logger.Error("Failed to encode data %s", err)
 		return
 	}
-	enc := json.NewEncoder(w)
-	defer w.Close()
-	if err := enc.Encode(&sam); err != nil {
-		a.logger.Error("Failed to encode file %s: %v", path, err)
+	if err = a.atomicallySaveToFile(data); err != nil {
+		a.logger.Error("Failed to encode file %s: %v", a.path, err)
 		return
 	}
 }
 
+func (a *AddrBook) atomicallySaveToFile(data []byte) error {
```

Comment:
why backup is needed? documentation for WriteFile says https://github.com/natefinch/atomic/blob/master/atomic.go#L12-L17

regarding crc check it is used to detect unexpected changes on disk (either disk malfunction, or someone messed with data accidentally). what makes sense to me:

on write:
- compute crc from json data
- write crc || json data

on read:
- decode crc from first 4(or 8) bytes
- compute another crc from the rest of the content
- compare both with each other
- if they are not equal - file can be discarded, as it got corrupted

Reply:
after write `crc || json data` to file, we can straightaway load data and check crc32
if error - then rollback to backup

Reply:
what does it solve?

Reply:
in case file was corrupted after save - we can restore backup, instead reset book

Reply:
corruption may happen not only during write, but at any point in time while file is stored on the disk. verifying that file content is consistent right after writing it doesn't do much at all

Reply:
should we give possibility to node rollback to previous version in case last file is corrupted?

for example: 

1. peers.json is broken
2. try peers.json.backup. if ok - use this
3. peers.json.backup is broken - reset book

Reply:
in my opinion backups don't add any additional safety in this case, as they are stored in the same place. if main file got corrupted, what are the chances that backup wasn't corrupted? we can store 100 backups for example and hope that atleast one of them will be in consistent state, but it don't improve fault tolerance that much 

in my understanding adding crc is useful not because we can eliminate such failure completely, but to speed up recovery if it occurs. without crc  node will have to connect with peers from potentially invalid book, but all records might be invalid in some subtle ways, which will cause it to spend significant amount of time removing them from the book.